### ISSUE TITLE
Add iBeacon capability to the Wear OS app

### ISSUE THREAD

**Issue by rpatel3001** [home-assistant/android#2183]
**Is your feature request related to a problem? Please describe.**

Room presence is a fun addition to a home automation setup, and the phone app beacon has been working well for me, but I sometimes leave my phone in a different room. Tracking a watch that's always on my wrist would be nice.

**Describe the solution you'd like**

An additional setting for the companion app to enable iBeacon advertisements, similar to the phone app.

**Describe alternatives you've considered, if any**

I'm going to tinker with writing my own app as the APIs for this don't seem too convoluted, but I've not done it before and it's not scalable.

**Additional context**

**Comment by dshokouhi** [comment]
All current phone sensors will be evaluated once the wear OS app has a sensor UI for enabling and adding settings.  Will come sometime after #2033

**Comment by rpatel3001** [comment]
Got it, thanks

**Comment by rpatel3001** [comment]
If anyone else is interested in this and lands at this issue, I have a working proof of concept at https://github.com/rpatel3001/BeaconWear

Credit to the good folks at @AltBeacon, this was forked from https://github.com/davidgyoung/android-beacon-library-reference-kotlin and modified with code from https://altbeacon.github.io/android-beacon-library/beacon-transmitter.html.

There's still a ton of cruft for scanning and absolutely no user interface, but I'll probably clean it up eventually.

**Comment by wifiuk** [comment]
be handy to have beacon capability on the home assistant wear os app for sure.

**Comment by maartendamen** [comment]
Would love to see this in comination with ESPresense. 
@rpatel3001 can you please explain a bit more about your PoC. Can I use it on my GW4 by just compiling and installing it?

**Comment by rpatel3001** [comment]
@maartendamen there is an apk available [here](https://github.com/rpatel3001/BeaconWear/releases/tag/first release). The UI doesn't actually do anything, the UUID and reference power are hard coded. You can clone the repo and build in Android Studio if you need to change those values.

**Comment by maartendamen** [comment]
> @maartendamen there is an apk available [here](https://github.com/rpatel3001/BeaconWear/releases/tag/first_release). The UI doesn't actually do anything, the UUID and reference power are hard coded. You can clone the repo and build in Android Studio if you need to change those values.

Very cool, I got it working. Thanks for your efforts!
I will do some writing about this on my blog soon if you don't mind.

**Comment by rpatel3001** [comment]
> Very cool, I got it working. Thanks for your efforts! I will do some writing about this on my blog soon if you don't mind.

Not at all, drop a link here when you do.

**Comment by ignacio82** [comment]
Any updates about this?

**Comment by dshokouhi** [comment]
> Any updates about this?

the wear OS app still needs to handle permissions and sensor settings before beacon capabilities can be added.

**Comment by marazmarci** [comment]
Idea: the transmitter could be enabled for a specific (short) amount of time, for eg. quickly detecting which room the user is in.

**Comment by Ulrar** [comment]
I don't see how you'd implement that, the watch won't know when you're changing room and if you're talking about enabling it only every few minutes it's going to make it a bit useless for automation.

I do like the idea of enabling it / disabling it through notifications though, you could have HA enable it when you're at home and disable it when away for example, neat but not essential IMHO

**Comment by marazmarci** [comment]
Sorry, I didn't explain my whole concept that I had in my head.
- It could be enabled triggered by motion sensors in the roooms.
- Some watches have multiple buttons. If the HA Wear OS app would have multiple launcher icons, that would trigger some event in HA, they could be used to eg. toggle the light in the current room you're in, by pressing the button on the watch, then enabling the BLE beacon for a short amount of time, then one of the beacon monitors would pick the signal up, and an automation can toggle the lights in the detected room that you're in.
(FYI there's an app called [Button Launcher](https://play.google.com/store/apps/details?id=com.brouken.wear.butcher) that lets you open more than one app or launcher item with a physical button. It could be useful for these kind of things.)

**Comment by Ulrar** [comment]
Not sure what the delay on detection might actually be but that's a neat idea. Sounds like the opposite might be better suited for this (having beacons broadcasting which room you're in and let the watch relay that over to HA) but that's a whole different feature.

Either way IMHO a first version always on / always off would be great, and it can be refined later with notification or events, but that's ultimately left to people who know how to make WearOS stuff :)

**Comment by dshokouhi** [comment]
> Idea: to not drain battery, the transmitter could be enabled for a specific (short) amount of time, for eg. quickly detecting which room the user is in.

how would that work? How would the transmitter know to turn itself on and off to make it useful to any automation when we cant send that data back to the watch to process (i.e. notification command on the watch)?

> I do like the idea of enabling it / disabling it through notifications though, you could have HA enable it when you're at home and disable it when away for example, neat but not essential IMHO

Thats what the phone app supports. The reason why its essential for a watch is because the only way you can enable and disable a sensor is to dig into settings on the watch, find the sensor and enable or disable it as you please. On a watch thats not  really ideal. When this sensor was added to the phone app the notification command was added at the same time because you need a good and reliable way to enable or disable it. Not having it automated defeats the purpose of a wearable here where its all about short interactions instead of needing to dig into places.



> * It could be enabled triggered by motion sensors in the roooms.

thats exactly what a notification command will allow you to do



> * Some watches have multiple buttons. If the HA Wear OS app would have multiple launcher icons, that would trigger some event in HA, they could be used to eg. toggle the light in the current room you're in, by pressing the button on the watch, then enabling the BLE beacon for a short amount of time, then one of the beacon monitors would pick the signal up, and an automation can toggle the lights in the detected room that you're in.
>   (FYI there's an app called [Button Launcher](https://play.google.com/store/apps/details?id=com.brouken.wear.butcher) that lets you open more than one app or launcher item with a physical button. It could be useful for these kind of things.)

I believe apps only have access to those buttons when the app is open.

https://developer.android.com/training/wearables/user-input/physical-buttons

The link to the button launcher is giving a 404



> Not sure what the delay on detection might actually be but that's a neat idea.

the delay may be there until the screen turns on the wearable due to doze restrictions



> Sounds like the opposite might be better suited for this (having beacons broadcasting which room you're in and let the watch relay that over to HA) but that's a whole different feature.

we have the opposite already on teh phone side, a beacon monitor sensor to report distance on nearby beacons. However this one also has a corresponding notification command that was introduced alongside the sensor and has the same battery considerations as the transmitter.

> Either way IMHO a first version always on / always off would be great, and it can be refined later with notification or events, but that's ultimately left to people who know how to make WearOS stuff :)

So if you can only enable and disable by physically interacting with the app how do you see that being useful? Are you goign to manually turn the sensor when you park your car and turn it off when you get in your house? Are you going to remember to turn it on and off before and after bed? What about when you leave the house?  The first version introduced in the phone side contained the notification command.

**Comment by Ulrar** [comment]
> So if you can only enable and disable by physically interacting with the app how do you see that being useful? Are you goign to manually turn the sensor when you park your car and turn it off when you get in your house? Are you going to remember to turn it on and off before and after bed? What about when you leave the house? The first version introduced in the phone side contained the notification command.

I'd just leave it on all the time, most likely.
The notification support would be great, just saying I don't see it as a blocker myself as long as it defaults off. But I understand the blocker now is the settings menu so this is moot

**Comment by dshokouhi** [comment]
> I'd just leave it on all the time, most likely.

Good to know I did not expect anyone to want to leave it on all the time lol

> The notification support would be great, just saying I don't see it as a blocker myself as long as it defaults off.

it defaults to on actually because you are enabling the sensor so the expectation is that the transmitter also gets enabled. The transmitter itself is controlled by a sensor setting which is not available in Wear OS yet, the notification command makes it easier to use and automate. 

We still need to do battery implication testing too, for example would you really want the sensor if your battery goes to 0 in 2 hours?

One thing I also considered was maybe adding a tile for turning the transmitter/monitor on and off, while it would make it faster it still wouldn't be automated.

**Comment by marazmarci** [comment]
> The link to the button launcher is giving a 404

Hmm, maybe it's because the developer didn't update it to comply with the latest `targetSdkVersion` requirements, and I can see it just because I have it installed. Btw here's the app's source code on GitHub: https://github.com/moneytoo/ButtonLauncher

**Comment by gsicilia82** [comment]
About battery drain:
I am using Xiaomi Mi Band 5 as "Beacon simulator". Sending ble advertisements is switched on all the time. It sends every 5 seconds. Battery capacity from Mi Band is small, but it holds for more than 10 days. Therefore, I don't expect to face battery drain on wear OS. It should have an impact, but not so heavy as discussed. Has anyone tested POC from @rpatel3001 and checked battery impact?

**Comment by Mausbiber** [comment]
I am trying to install the apk from @rpatel3001.
When I then try to start the app on my smartphone, I always get an error message:

But I can't find a way to give the app the desired permission.

Oh yes, I use a Pixel 7 Pro in conjunction with the Pixel Watch.

**Comment by rpatel3001** [comment]
@Mausbiber there's no error message in your comment, but it doesn't matter because the apk is meant to be installed on the watch not the phone. I can't say if the pixel watch will work, I've only tried on the galaxy watch 4.

**Comment by Mausbiber** [comment]
okay, trying to sideload

**Comment by Mausbiber** [comment]
Okay, I installed the app on the Pixel Watch via adb.
I can also start the app and the "useless" GUI appears.
But that's it, I'm not asked for rights and no iBeacon is sent.

**Comment by dshokouhi** [comment]
Hey guys as this app you are installing is not the HA android app I am going to ask you to please move this conversation to the appropriate location.

**Comment by soloam** [comment]
This would be a great addition, nothing better to make tracking that a watch!

**Comment by garv3** [comment]
Still waiting!

**Comment by Scags104** [comment]
> @maartendamen there is an apk available [here](https://github.com/rpatel3001/BeaconWear/releases/tag/first_release). The UI doesn't actually do anything, the UUID and reference power are hard coded. You can clone the repo and build in Android Studio if you need to change those values.

This saved me from returning my 300$ pixel watch. thank you. this is awesome.

**Comment by dshokouhi** [comment]
Hey everyone I submitted a draft PR to migrate over the sensors to get some early feedback. I see a lot of users are coming in this thread looking for a quick solution. As of now there are no settings or notification commands but the transmitter and monitor should function just like the phone app does. If possible please test and provide us some feedback regarding your usage and battery life and also if you feel that the other options are a necessity for your use case. Its important for us to get some feedback here to see if things are good as it stands or if we really need to take care of other considerations first.

You can grab the test debug wear APK from the below link, just download the archive and extract it:

https://github.com/home-assistant/android/actions/runs/3744776429

Note you may need to install the debug phone APK in order to login on the watch, both APKs can be found in the archive from above.

once done with the test  you can leave it installed or uninstall it, the debug builds are designed to be installed side by side the production apps so they dont interfere with them.

**Comment by Ulrar** [comment]
Hey @dshokouhi,
that's great news, thank you very much !

I've just installed it now and I'll test it over the next while, I'm going to need to setup presence tracking here to see if it works but I'll be off for a few days there so I'll work on that.
For anyone else wanting to try this, you need to install the 'full' app on the phone, looks like the minimal one doesn't support wearos.

I'll update back once I've had a chance to test it out properly, thanks again !

**Comment by soloam** [comment]
Thank you @dshokouhi . Will you push this to the beta?

Tks

**Comment by Ulrar** [comment]
Battery drain doesn't look noticeably worst so far, but straight away I can already see what I imagine is a bug : 
![image](https://user-images.githubusercontent.com/873423/208915859-fbd103ac-3814-4d8a-9ea6-f6c20eded339.png)

It looks like it's regularly being picked up as a new device by HA, making it essentially not trackable it seems

**Comment by dshokouhi** [comment]
> It looks like it's regularly being picked up as a new device by HA, making it essentially not trackable it seems

That's to be expected. You have to wait until it has 10 entries then HA core will combine then.

**Comment by dshokouhi** [comment]
> Will you push this to the beta?


Not until after testing

**Comment by dshokouhi** [comment]
> Battery drain doesn't look noticeably worst so far

It's not draining the watch immediately but it is draining faster with it enabled.

**Comment by Ulrar** [comment]
So after a couple of weeks, it looks like the debug version of the app uses about 0.2% of the battery, whereas the normal app (that I actually use with sensors and stuff) uses about 0.4. So I suppose we may be looking at an extra 0.2, which I personally haven't noticed on daily use.

As for the functionality itself it's been spotty for me, but then again the ESPHome bluetooth_proxy has been having a lot of issues since 2022.12 so it's probably not the app's fault. If the concern here for this preview was the battery use, IMHO, it's all good !

**Comment by dshokouhi** [comment]
@Ulrar just to confirm did you have just the transmitter sensor enabled or also the beacon monitor?

**Comment by Ulrar** [comment]
Just the BLE transmitor, that shows up as an ibeacon in HA.
(BTW it did merge after a while, all good. I've never used one before so I didn't know)

**Comment by rpatel3001** [comment]
I don't have battery numbers as I've just pulled mine off the charger but I'll try and remember to look before charging next. It doesn't seem to have had a large affect, and I'm sure that adding the settings for lower power/less frequent broadcasts would give an even lower impact.

**Comment by dshokouhi** [comment]
I think its a positive thing that nobody has came here to say the watch died over night when they went to bed with the battery at 100%

is everyones intention then to leave it on all the time? Would a tile to control the state of the transmitter be a good idea or do you think it woudl not get used?

**Comment by rpatel3001** [comment]
A way to enable/disable transmitting through automations, like the phone app transmitter, would be ideal.

**Comment by dshokouhi** [comment]
yea that will come with notifications once the watch gains support for them directly, that would actually negate the need to control settings on the watch as we have commands to control all the settings for this sensor :)

That is actually a separate feature request 😁

**Comment by soloam** [comment]
Since it's not possible to turn on and off by notification, I would leave it on all day long!

**Comment by dshokouhi** [comment]
Thanks for the feedback everyone! I went ahead and marked my PR as ready for review. Also updated the docs to mention sensor settings are not currently supported yet so all we can do is enable or disable these sensors.

**Comment by rpatel3001** [comment]
Another battery data point: after 48 hours my watch drained from 100% to 5%. The blue app without BLE reports 0.5% usage and the red app with transmitting enabled the whole time reports 0.7% usage.

**Comment by Scags104** [comment]
installed latest betas with this merged. is there a way to change the transmit power from ultra low that im not aware of? Watch app doesn't seem to have the settings behind each sensor like the phone app?

Not seeing it broadcasting as an iBeacon with MQTT explorer. 

Thanks

**Comment by dshokouhi** [comment]
@Scags104 it's been mentioned a few times in this issue that sensor settings do not work on wear os. Also mentioned in the docs.

**Comment by soloam** [comment]
@dshokouhi is this in Beta? I'm a beta user and still didn't received the update!

Thank you for all your work on this

**Comment by dshokouhi** [comment]
Beta gets pushed every Saturday night to play store then waits for review before being available. You can grab the APK off the GitHub release page and install it now.

**Comment by Mausbiber** [comment]
Where did i find the apk?

**Comment by dshokouhi** [comment]
@Mausbiber github releases page.

https://github.com/home-assistant/android/releases

**Comment by Mausbiber** [comment]
Where is the difference between full, minimal und wear release?

**Comment by dshokouhi** [comment]
> Where is the difference between full, minimal und wear release?

Its mentioned in the project read me and in the documentation.

https://github.com/home-assistant/android#app-flavors
https://companion.home-assistant.io/docs/core/android-flavors

Full version is Google Play store, minimal has no play services meaning no Wear OS.

If you have any more generic questions please hop on discord or the forums.

**Comment by Mausbiber** [comment]
I downloaded the full-release beta and used it to update the HA app on my smartphone.
However, I don't see an iBeacon sensor or the option to turn it on in the sensor settings on the Pixel Watch.

**Comment by dshokouhi** [comment]
@Mausbiber you need to update the watch and not just the phone.  Please hop on Discord if you continue to have issues.

**Comment by Mausbiber** [comment]
which discord channel?

**Comment by mmitch2k** [comment]
sorry im still a little confused. are you saying that the Home Assistant app thats in the Play store (which i have installed on my pixel watch) provides the ibeacon capability or do I need to download something else. I dont see that option anywhere in the app

**Comment by dshokouhi** [comment]
Just a quick update to everyone following this thread. The notification commands for BLE transmitter and Beacon Monitor are now added to Wear OS starting from Beta 1114. This is not yet on the play store but you can always sideload from the github releases tab. Docs are not up to date yet but will be soon, they use the same format as the phone command so you can refer to those commands.

**Comment by mmitch2k** [comment]
> Thank you i now have the watch reporting with espresence on high and low latency.
> 
> Still have one issue non-related to this issue but hopefully there is a fix. seems the bluetooth connection between the phone and pixel watch is causing a broadcast issue, if i toggle bluetooth from the phone it works for about a minute before messages stop sending.
> 
> appreciate the continued work on this!

Hi is the watch now showing up in mqtt explorer for you?

**Comment by Scags104** [comment]
@mmitch2k yes but only 2-3 messages before , and im guessing here, the connection between my phone and watch kills the transmit abilities. more messages come through when i toggle my phones bluetooth connection. very short lived.

**Comment by mmitch2k** [comment]
@Scags104 how do you even get it to 2-3 messages. i have the latest beta, is there a setting in there that im missing? Because for the life of me I cant get it to work. I am at the point now where i have an external becaon attached to my watch :)

**Comment by Scags104** [comment]
@mmitch2k send messages to watch from HA following the documentation [here](https://companion.home-assistant.io/docs/notifications/notification-commands/#ble-beacon-transmitter) to change transmit power to high and advertise mode to low_latency

**Comment by mmitch2k** [comment]
@Scags104 yup same issue here. stops broadcasting after a few minutes

**Comment by dshokouhi** [comment]
For those seeing the transmitter stopping, check if you are able to grant the app background access. I know this is possible on Samsung watches through their companion app. If you still experience an issue you can file a new issue and grab the logs. This issue may not be one we can solve as Wear OS does not offer a way for apps to have background access. They are susceptible to doze and background limits will vary from device to device.

**Comment by Scags104** [comment]
@dshokouhi unfortunately not seeing that as an option for the pixel watch. Anything you can think of in terms of sending a message, given the new functionality, at a specified frequency via automation to force-wake it?

**Comment by dshokouhi** [comment]
@Scags104 try using the beacon monitor on the phone to see if it still sees the transmitter when you see that issue. I am not really sure at what interval it stops transmitting, if it encountered an error I would imagine the state of the sensor should update. We would need to see the logs when the issue occurs to see what we can do.

**Comment by Scags104** [comment]
@dshokouhi i still see it in the attributes of the monitor but unclear where to enable logging specifically to assist (phone or HA instance etc). 

Side note i tried an automation every 30s to update sensors on the mobile app on the watch, as well as sending a notification to update the transmitter latency. neither worked to wake the watch as i was staring at mqtt explorer and only seeing the same initial 3 messages which had been broadcast - that idea is out.

**Comment by dshokouhi** [comment]
> @dshokouhi i still see it in the attributes of the monitor but unclear where to enable logging specifically to assist (phone or HA instance etc).

then this suggests that the transmitter is indeed still transmitting so the issue probably lies on the other end detecting the beacon. The transmitter and monitor use the same library so its the best test to see where the issue lies.

> Side note i tried an automation every 30s to update sensors on the mobile app on the watch,

This command has not been added to Wear OS yet

>  as well as sending a notification to update the transmitter latency. neither worked to wake the watch as i was staring at mqtt explorer and only seeing the same initial 3 messages which had been broadcast - that idea is out.

did you make sure to use the critical notification format to actually wake up the watch? Same rules apply to notifications on a watch like from the phone, FCM always tries to respect doze unless we force it.

Only the first example applies to Wear OS

https://companion.home-assistant.io/docs/notifications/critical-notifications#android

**Comment by Scags104** [comment]
> This command has not been added to Wear OS yet
> 
 i just meant i created an automation using HAs service to update sensors. I also tried to use the notification service to update command: "ble_set_advertise_mode" settings every few seconds which didnt wake as intended either. I did try it also with the `ttl: 0
priority: high` 
to no avail.

**Comment by dshokouhi** [comment]
> i just meant i created an automation using HAs service to update sensors.

you mean `home_assistant.update_entity` ? that wont work either :)



> I also tried to use the notification service to update command: "ble_set_advertise_mode" settings every few seconds which didnt wake as intended either. I did try it also with the `ttl: 0 priority: high`
> to no avail.

we would need to see the on device logs to see if there is an actual delay but that is not related to this issue :)

**Comment by Scags104** [comment]
@dshokouhi  thanks i split this out to its own issue [3272](https://github.com/home-assistant/android/issues/3272)

**Comment by dshokouhi** [comment]
its actually a duplicate of https://github.com/home-assistant/android/issues/3159 so please add further comments in there and you can see what that user did to fix their issue. Given that the beacon monitor picks up the value the transmitter is not stopping, if it did the monitor would also not report it.
